### DISCUSSION TITLE
About using/implementing multiple hidden layers in CudaLSTM

### DISCUSSION THREAD

**Discussion by jpcurbelo** [neuralhydrology/neuralhydrology#170] [Q&A] [answered]
First off, I want to extend my thanks for the outstanding work you've done and for generously sharing this tool.

I'm currently working on generating benchmark results for a new dataset covering USA and CAN basins. One of my initial experiments involves a multi-basin model focusing solely on dynamic variables, essentially an 'LSTM without static inputs', utilizing the CudaLSTM model. However, I'm interested in exploring the potential benefits of incorporating multiple hidden layers into the LSTM architecture.

I'm considering two options for this:

1 - Updating CudaLSTM model: One possibility is to update the existing CudaLSTM code to support multiple hidden layers. I'm wondering if this is achievable and if so, whether I could contribute by submitting a PR for the necessary modifications.

2 - Using EA-LSTM model: Alternatively, I could explore using the EA-LSTM model and adjust the configuration file to exclude static variables.

I'd appreciate any insights or advice you might have on the best approach to take.

Thank you once again for your exceptional work and for fostering collaboration in the community.

**Comment by kratzert** [comment, marked as answer]
Re 1. Simple pass `num_layers=n`, where `n` is the number of stacked LSTM layers that you want to use, to https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/modelzoo/cudalstm.py#L37, see https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#lstm We discussed this in the past @gauchm, right? But IIRC we never saw the need to make this an implemented config argument, since it is such an easy change for anyone who wants to test it.

Re 2: The EA-LSTM is not defined without static inputs.

**Reply by jpcurbelo** [reply]
Thanks for your prompt response! I have a couple of questions about the experiments on the papers and the metrics presented, not directly related to this ticket though. I thought of opening a new ticket for those. Would that be the proper channel for those?

**Reply by gauchm** [reply]
Yeah, like you (@kratzert) say, it's easy to hack in when needed, so we never bothered. But I guess it's a low-hanging fruit -- I wouldn't be opposed to merging it if someone likes to implement it and sends a PR (bonus points for making it general, also supporting GRU and maybe even CustomLSTM and others) :)
